protocol/contracts/amo/TpfAmo.sol

```diff
@@ -0,0 +1,549 @@
+        uint256 templeAmountIn,
+        uint256 minBptOut
+    ) external onlyOperatorOrOwner whenNotPaused enoughCooldown {
+        _validateParams(minBptOut, templeAmountIn);
+        
+        // check spot price is above TPF
+        if (!poolHelper.isSpotPriceAboveTPFUpperBound()) {
+            revert AMOCommon.NoRebalanceDown();
+        }
+
+        // will join take price below tpf with lower bound
+        // should rarely be the case, but a sanity check nonetheless
+        if (poolHelper.willJoinTakePriceBelowTPFLowerBound(templeAmountIn)) {
+            revert AMOCommon.HighSlippage();
+        }
+
+        AMO__ITempleERC20Token(address(temple)).mint(address(this), templeAmountIn);
+
+        // create request
+        AMO__IBalancerVault.JoinPoolRequest memory joinPoolRequest;
+        if (templeBalancerPoolIndex == 0) {
+            joinPoolRequest = poolHelper.createPoolJoinRequest(temple, stable, templeAmountIn, 0, minBptOut);
+        } else {
+            joinPoolRequest = poolHelper.createPoolJoinRequest(temple, stable, templeAmountIn, 1, minBptOut);
+        }
+
+        // approve
+        temple.safeIncreaseAllowance(address(balancerVault), templeAmountIn);
+
+        // execute and sanity check
+        uint256 bptAmountBefore = bptToken.balanceOf(address(this));
+        uint256 spotPriceScaledBefore = poolHelper.getSpotPriceScaled();
+        balancerVault.joinPool(balancerPoolId, address(this), address(this), joinPoolRequest);
+        uint256 bptAmountAfter = bptToken.balanceOf(address(this));
+        uint256 bptIn;
+        unchecked {
+            bptIn = bptAmountAfter - bptAmountBefore;
+        }
+        if (bptIn < minBptOut) {
+            revert AMOCommon.InsufficientAmountOutPostcall(bptAmountBefore + minBptOut, bptAmountAfter);
+        }
+
+        // revert if high slippage after pool join
+        if (uint64(poolHelper.getSlippage(spotPriceScaledBefore)) > postRebalanceSlippage) {
+            revert AMOCommon.HighSlippage();
+        }
+        lastRebalanceTimeSecs = uint64(block.timestamp);
+        emit RebalanceDown(templeAmountIn, bptIn);
+
+        // deposit and stake BPT
+        // bptToken.safeIncreaseAllowance(address(amoStaking), bptIn);
+        bptToken.safeTransfer(address(amoStaking), bptIn);
+        amoStaking.depositAndStake(bptIn);
+    }
+
+    /**
+     * @notice Single-side deposit stable tokens into balancer pool when TEMPLE price 
+     * is below Treasury Price Floor.
+     * @param amountIn Amount of stable tokens to deposit into balancer pool
+     * @param minBptOut Minimum amount of BPT tokens expected to receive
+     */
+    function depositStable(
+        uint256 amountIn,
+        uint256 minBptOut
+    ) external onlyOwner whenNotPaused {
+        _validateParams(minBptOut, amountIn);
+
+        // check spot price < TPF
+        if(!poolHelper.isSpotPriceBelowTPFLowerBound()) {
+            revert AMOCommon.NoRebalanceUp();
+        }
+
+        // create request
```

Comment:
Can't we build the request off-chain and pass as an arg (like you do for add/remove liquidity)?

Reply:
did this to save on gas. rebalance functions are called way more often and the pool requests structs consume quite an amount of gas. Instead I just build the request using some `mstore` operations and only pass as arguments the necessary arguments

Reply:
Just to say back what I think you are saying: passing the pool request as a parameter (where you build it off-chain) consumes more gas than constructing it within the function?

Reply:
Yes. like i said, call with join or exit request contains a lot of bytes that can be trimmed off.
Also, the request will be passed on to the `poolHelper`, further using up more gas.

Here's further proof on first and next calls using pool request vs no pool request :

```
function callWithRequest(
        JoinPoolRequest memory request,
        uint256 amountOut
    ) external {
        ans = 10;
        // first call: Execution cost 56401 gas 
        // next calls: Execution cost 36501 gas
    }

    function callNoRequest(
        uint256 amountIn,
        uint256 amountOut
    ) external {
        address[] memory assets = new address[](2);
        uint256[] memory maxAmountsIn = new uint256[](2);
    
        JoinPoolRequest memory request;
        (assets[0], assets[1]) = index == 0 ? (temple, stable) : (stable, temple);
        (maxAmountsIn[0], maxAmountsIn[1]) = index == uint256(0) ? (amountIn, uint256(0)) : (uint256(0), amountIn);

        request.userData = abi.encode(uint256(1), request.maxAmountsIn, amountOut);
        request.assets = assets;
        request.maxAmountsIn = maxAmountsIn;
        
        ans = 10;
        // firs call: Execution cost 52065 gas
        // next calls: Execution cost 32165 gas
    }
```